### PR TITLE
feat(FR-4040): connect a custom OpenAI-compatible endpoint from the Chat page

### LINKED ISSUE #9876: Connect a custom OpenAI-compatible endpoint from the Chat page
## Background

The Chat page can only talk to a Backend.AI deployment. Pasting an arbitrary OpenAI-compatible URL used to work by selecting a deployment whose `/models` call failed and typing an absolute URL into the recovery form's Base Path field. FR-3332 narrowed the deployment selector to deployments with active replicas, so that state is almost unreachable now, and the Astryx port dropped the Base Path input's URL prefix that hinted at it. A user reported the loss on Teams (UX Ideas channel, 2026-09-21).

Registering and managing custom endpoints will come with AppConfig in the LTS release. This issue covers the simple form for the current release: connect a custom endpoint directly from Chat, without saving it anywhere but the chat history.

## Scope

- The deployment selector's panel footer offers **Connect a custom endpoint…**, also when no deployment is listed. It opens an inline form for a base URL and an optional API key.
- **Connect** fetches `/models` from inside the form and applies the endpoint only when at least one model comes back. A failure stays on the form with a cause-specific message: 401/403 points at the key, 404 suggests the usual `/v1` suffix with a one-click retry, network failures mention CORS (the browser calls the endpoint directly), plus timeout, 5xx, non-OpenAI responses and an empty model list.
- While a custom endpoint is selected, the selector shows its host and the detail-page button becomes an edit button that reopens the form.
- The API key is kept per chat panel in `sessionStorage`, never in the chat history (`localStorage`). Two panels may use the same endpoint with different keys. A history entry restored without its key asks for it again; cloning a panel for comparison copies the key.
- `?baseURL=` opens the custom form as a deep link; the key is never read from the URL.

## Out of scope

- Saving or managing custom endpoints (LTS, AppConfig).
- A site policy to disable custom endpoints.

JIRA Issue: FR-4040

### PR DESCRIPTION
**Pull request by yomybaby**
Resolves #9876 (FR-4040)

## Summary

Chat can now talk to any OpenAI-compatible endpoint without a Backend.AI deployment. The deployment selector's panel footer offers **Connect a custom endpoint…** (also when no deployment is listed), which opens an inline form for a base URL and an optional API key.

Before this, the only way to reach an arbitrary URL was to select a deployment whose `/models` call failed and paste an absolute URL into the recovery form's Base Path field. FR-3332 narrowed the selector to deployments with active replicas, which made that state almost unreachable, and the Astryx port dropped the Base Path input's URL prefix that hinted at it. Registering and managing custom endpoints is a separate, later change (AppConfig, LTS); this PR is the simple form for the current release.

## How it works

- **Connect** fetches `/models` from inside the form and applies the endpoint only when at least one model comes back. A failure stays on the form with a cause-specific message: 401/403 points at the key; 404 suggests the usual `/v1` suffix with a one-click **Retry with /v1**; a network failure mentions CORS, since the browser calls the endpoint directly; timeout, 5xx, a non-OpenAI response and an empty model list each get their own line.
- While a custom endpoint is selected, the selector shows its host and the detail-page button beside it becomes an **edit** button that reopens the form.
- The footer is ruled off from the option list. `BAIComplexSelect.footer` may now be a function of `close`, so a footer action can dismiss the panel.
- `?baseURL=` opens the custom form as a deep link. The "Compare with other models" menu item is hidden for a custom endpoint, since the deployments chatting tab addresses a deployment.
- The old Base Path route still works; nothing was removed.

## API key handling

The key never enters the chat history. The persisted provider record (`localStorage`) holds the URL only; the key lives in `sessionStorage` keyed by **chat panel id** (`customEndpointKeyStore.ts`), so two panels can use the same endpoint with different keys, a reload keeps it, and closing the tab drops it. A history entry restored without its key gets the form back with an "enter the API key again" banner. Cloning a panel for comparison copies the key to the new panel. The key is never read from the URL.

## Changes

- `react/src/components/Chat/CustomEndpointForm.tsx` (new) — the inline form, in-form `/models` probe, error guidance.
- `react/src/components/Chat/customEndpointKeyStore.ts` (new, + test) — per-panel session key store.
- `react/src/components/Chat/openAIModels.ts` (new) — shared `/models` fetcher with classified errors; `ChatCard`'s `useModels` uses it too.
- `DeploymentSelect.tsx` — footer entry, custom label, edit button. `ChatHeader.tsx`, `ChatCard.tsx`, `ChatModel.tsx`, `ChatHistory.ts`, `ChatPage.tsx` — wiring, `baseURL` provider field, deep link, key copy on clone.
- `packages/backend.ai-ui/src/components/BAIComplexSelect.tsx` / `.css` — `footer` as a function of `close`; `bai-complex-select__foot--divided`.
- i18n `chatui.customEndpoint.*` (en, ko).

## Verification

- `bash scripts/verify.sh` — `=== ALL PASS ===`
- `node scripts/migration-gates/astryx-token-gate.mjs --strict` — no new findings
- `vitest` — `customEndpointKeyStore.test.ts` 8/8, Chat suite 15/15, `BAIComplexSelect` 18/18
- Live check pending: the test manager (`10.82.0.130`) was unreachable while this was built.

## Test server

`http://10.82.0.130:8090`

### CODE REVIEW

**Comment by yomybaby** [comment]
Dev server on box `jongeun`: http://fr-4040-pr9877-connect-custom-openai.jongeun.10-82-0-159.sslip.io

File: react/src/components/Chat/customEndpointKeyStore.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    return {};
+  }
+}
+
+let snapshot: KeyMap = read();
+
+function write(next: KeyMap) {
+  snapshot = next;
+  try {
+    sessionStorage.setItem(STORAGE_KEY, JSON.stringify(next));
```
[the remark this answers is not in the document]
Comment (yomybaby) [reply react/src/components/Chat/customEndpointKeyStore.ts]: Fixed in c9966de63: the custom endpoint API key is no longer written to `sessionStorage`. `customEndpointKeyStore.ts` now keeps it only in a module-level map keyed by chat panel id, so nothing reaches Web Storage and a page reload asks for the key again. Per-panel keys and the copy on "Add comparison chat" are unchanged; the test now asserts that `sessionStorage` and `localStorage` stay empty after a key is set.

**Review by agatha197 — CHANGES_REQUESTED** [review]
Please translate other languages.

**Comment by yomybaby** [comment]
Addressed @agatha197's review: translated `chatui.customEndpoint.*` into all 19 remaining locales in `21a31f086` (de, el, es, fi, fr, id, it, ja, mn, ms, pl, pt, pt-BR, ru, th, tr, vi, zh-CN, zh-TW), reusing each locale's existing terms for endpoint and API key. Every locale now carries the same 23 keys as `en.json`; `Base URL`, `/v1`, `/models` and `CORS` stay literal. Also dropped the unused `ConnectionFailed` key from en/ko. `verify.sh` ALL PASS.

**Comment by agatha197** [comment]
## Live check: custom endpoint against an open model (Ollama)

Tested this PR's head (21a31f0) from a local dev server on the `10.82.130.172` backend, with a local Ollama serving `qwen2.5:0.5b` at `http://127.0.0.1:11434/v1` (`OLLAMA_ORIGINS=*`).

**Passed**
- Deployment selector → **Connect a custom endpoint…** → Connect: `GET /v1/models` 200, selector shows `127.0.0.1:11434`, model auto-selects `qwen2.5:0.5b`, the detail button becomes **Edit custom endpoint**.
- Sending a message: `POST /v1/chat/completions` 200, streamed reply renders, TPS/token counter updates, Stop returns to Send.
- Reload on the same chat URL restores the endpoint, model and messages; a second turn works.
- Base URL without `/v1` shows the 404 banner with **Retry with /v1**, and the retry reconnects.
- With an API key, `/models` and `chat/completions` both carry `Authorization: Bearer …`; the key is absent from localStorage/sessionStorage after the chat is saved.

| Chat reply from the open model | After reload | Missing `/v1` guidance |
|---|---|---|
| reply | reload |  |

**One observation (non-blocking):** after a reload without a key, the requests go out with an empty `Authorization:` header (`apiKey ? `Bearer …` : ''` in `openAIModels.ts`). Ollama accepts it, but some servers reject an empty Authorization header, so omitting the header when there is no key would be safer.

**Review by agatha197 — APPROVED** [review]
LGTM
